Refactor evidence generation and validation

tskir opened the pull request:
Main ticket: https://github.com/opentargets/platform/issues/1834

**Fix bugs and issues**
- [x] Fix paths in the configuration
- [x] Fix compile critical syntax errors in Snakefile
- [x] Fix EPMC rule input
- [x] Fix CancerBiomarkers rule input
- [x] Fix Java dependency installation
- [x] Fix specifying the PYTHONPATH
- [x] Fix detecting Spark memory limit
- [x] Fix Orphanet rule syntax
- [x] Fix recursive FTP download
- [x] Fix NoneType replace error in _ontoma_udf
- [x] Fix and simplify EPMC fetch
- [x] Fix Java heap space error for EPMC
- [x] Fix Java heap space error for Gene2Phenotype
- [x] Fix unmappedDiseases output for Gene2Phenotype
- [x] Fix Orphanet output type to be a file
- [x] Fix retries in HTTP RemoteProvider
- [x] Fix Snakemake params formatting

**Data changes**
- [x] Use the most recent drugIndex dataset
- [x] Use OnToma v1.0.1 and the latest available EFO version

**Refactor**
- [x] Unify coalescing and writing the evidence strings
- [x] Consistently sort the parsers in the code
- [x] Simplify Snakefile by removing fetch* steps

**Implement validation**
- [x] Add opentargets-validator and update the environments
- [x] Add validation step to all Snakemake rules
- [x] Fix EPMC schema validation error
- [x] Fix EPMC PMID whitespace error
- [x] Fix three schema validation errors for EPMC, PanelApp and PhenoDigm

**Add and clarify instructions**
- [x] Add reproducible instructions on data processing in the cloud
- [x] Add instructions on using the service account
- [x] Add note on generating the Genetics Portal evidence

**Remove obsolete code and documentation**
- [x] Remove processing for the Genetics Portal from Snakefile
- [x] Remove obsolete information from README
- [x] Remove the two unused modules: RareDiseasesUtils, Utils

tskir commented:
**Pull request updated:** added note on processing Genetics Portal data. I moved it from the Snakefile into a separate portion of README, where it fits more appropriately

tskir commented:
@DSuveges @ireneisdoomed PR is now significantly expanded and now includes additional bug fixes, validation, and accompanying refactoring. The description is also updated. I'm currently running evidence generation with this latest version, fingers crossed

ireneisdoomed reviewed (commented):
Thank you for all the significant changes, it looks super clean.
I have a question: the output of the parser is used in the validation step. Do you know if snakemake is downloading the remote file to feed it into the validator?
If this is the case and to avoid extra downloading steps, I'd suggest storing the parsers' output locally, then validating it and then copying it to the remote location.

Another thing, I'd suggest storing `write_evidence_strings` in a different module other than `spark.py`. My suggestion is to create `evidence.py` to store the common functions used across the generation steps.
Alternatively, if we want to avoid having many modules, `utils.py` could act as a factory where these snippets live and modularise it in a later stage if we eventually have many evidence related functions.

ireneisdoomed commented:
> If this is the case and to avoid extra downloading steps, I'd suggest storing the parsers' output locally, then validating it and then copying it to the remote location.

I didn't know the output was deleted if the whole rule failed, perhaps this does not hold anymore.

DSuveges reviewed (changes requested):
Everything really makes sense. My only comment that needs to be addressed is that the env file needs to be updated so the OT validator is installed so the evidence validation runs.

tskir commented on modules/EPMC.py:
```diff
@@ -58,8 +51,11 @@ def main(cooccurrenceFile, outputFile, local=False):
+
         # Casting integer pmid column to string:
-        .withColumn("pmid", pf.col('pmid').cast(StringType()))
+        .withColumn("pmid", pf.trim(pf.col('pmid').cast(StringType())))
```
That's a good question! The story of this `trim()` call is as follows. When I was originally running this, validation of the EPMC results failed because _one record_ contained an extra space symbol in its `pmid` field. This _probably_ (I didn't check, because I really didn't have time in between fixing all of the other bugs :-) caused Spark to either load the entire column as strings, or just that one record.

So, in sort, it's not that `cast(StringType())` does anything nasty, it's just to guard against similar issues in the original data.

DSuveges replied:
It's interesting... I vaguely remember adding this statement as the evidence was failing validation as all pmids were integers. I guess at some point the problematic pmid got introduced and since then, spark probably won't cast that column into integer. Good. Thanks for clarifying.

tskir commented on README.md:
```diff
@@ -1,60 +1,94 @@
+# Activate the environment.
+git clone https://github.com/opentargets/evidence_datasource_parsers
+cd evidence_datasource_parsers
+conda env remove -n evidence_datasource_parsers_lock
```
But doesn't line 39 (`exec bash`) do just that? Or did that not work for some reason?

https://github.com/opentargets/evidence_datasource_parsers/pull/103/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R39

DSuveges replied:
Yeah, I must have missed about that line:
```
dsuveges@evidence-generation-test:~$ exec bash
(base) dsuveges@evidence-generation-test:~$ which conda
/home/dsuveges/miniconda3/bin/conda
```

tskir commented on README.md:
```diff
@@ -1,60 +1,94 @@
+git clone https://github.com/opentargets/evidence_datasource_parsers
+cd evidence_datasource_parsers
+conda env remove -n evidence_datasource_parsers_lock
+conda env create --file envs/environment-lock.yml
```
It is though, line 165: https://github.com/opentargets/evidence_datasource_parsers/pull/103/files#diff-19bd9546d25e5bd92de1d9812a81e2cfa2aefb5ffdcf0f1ceb6f9575d921d81aR165

It's very strange if it doesn't work, could you please check that you have the latest version of this PR checked out?

DSuveges replied:
Oh, that's strange... I didn't see any errors when building the environment, but all the validation step failed due to `opentargets_validator` not being in the path. Unfortunately I have destroyed the VM, so all logs got lost. (all log files under `gs://otar000-evidence_input/parser_logs/` are empty.)

I give it an other go. I'll watch out for the other comment.

DSuveges replied:
OK, I got it: I built the environment from the master branch. I did try to run the snakemake from the master, but that failed and then switched to your branch, however I didn't rebuilt the env. 

When doing things right, the individual parsers run and validates. However, when running all rules I get this:
```
(evidence_datasource_parsers_lock) dsuveges@evidence-generation-test:~/evidence_datasource_parsers$ snakemake --cores all
The flag 'directory' used in rule epmc is only valid for outputs, not inputs.
Building DAG of jobs...
MissingInputException in line 17 of /home/dsuveges/evidence_datasource_parsers/Snakefile:
Missing input files for rule all:
otar000-evidence_input/ClinGen/json/clingen-2021-11-17.json.gz
otar000-evidence_input/Orphanet/json/orphanet-2021-11-17.json.gz
otar001-core/MousePhenotypes/mouse_phenotypes-2021-11-17.json.gz
```
I couldn't find out what this mean exactly, when running these parsers they work just fine.

DSuveges replied:
I think something happened in this commit: 0cf2c9723f83edf1c7768b83107663930733e090

tskir replied:
Fixed now

tskir commented:
@ireneisdoomed @DSuveges Thank you very much for the reviews. I've addressed the discussions in-place, and now for the remaining questions:

> I have a question: the output of the parser is used in the validation step. Do you know if snakemake is downloading the remote file to feed it into the validator?
> If this is the case and to avoid extra downloading steps, I'd suggest storing the parsers' output locally, then validating it and then copying it to the remote location.

That's a good point, but thankfully Snakemake handles this case properly. The remote input/output files are only transferred once all operations on them have finished. So there's no extra downloading happening.

> I didn't know the output was deleted if the whole rule failed, perhaps this does not hold anymore.

Yes, that's also correct, although as I understand it, it's more of a separate thing, to ensure that there aren't any corrupted/half written files (either locally or remotely).

> Another thing, I'd suggest storing write_evidence_strings in a different module other than spark.py. My suggestion is to create evidence.py to store the common functions used across the generation steps.
> Alternatively, if we want to avoid having many modules, utils.py could act as a factory where these snippets live and modularise it in a later stage if we eventually have many evidence related functions.

Sure, let's do it this way then, I don't mind. I've renamed `spark.py` to `evidence.py`.
